### ISSUE TITLE
Security Issue: Unrestricted File Upload

### ISSUE THREAD

**Issue by narendrabhati** [wolfcms/wolfcms#625]
Hello There Wolf CMS ! Greeting From Bhati

While looking at your cms version 0.8.2 i found that, an authenticated user can upload a malicious arbitrary file to server which allow him to take over the web server access like command execution where an user can act as server admin.

Looking forward for your response so i can share the details with you ! it will better if we discuss in email so the information will not disclose here !

**Comment by mvdkleijn** [comment]
this is almost starting to feel like spam... a set(!) of indian guys reporting the exact same "vulnerability" in almost the same sentence structure. No offense intended by the way.

You can send you info to wolfcms at gmail dot com

**Comment by narendrabhati** [comment]
Shocked to see your response , 
1. My replied on comment about "Indian Researcher Spamming On You"
First you have to take your words back that indians are spamming on you, Facebook has announced that India Tops Facebook's Bug Bounty Program Again With Most Recipients.
http://gadgets.ndtv.com/social-networking/news/india-tops-facebooks-bug-bounty-program-again-with-most-recipients-664843

2. Report Structure Is Same

Report structure will always be the same due to vulnerability details , If a person will report you from India , USA ,  London then still the structure will be the same , So think before the reply

3. You can send you info to wolfcms at gmail dot com
No no no , I am very sorry that i notify you about your cms vulnerability , I have 3 more critical vulnerabilities but i will not report ! why ?
Because you are feeling that Indians are spamming on you !

**Comment by mvdkleijn** [comment]
I apologize for the misunderstanding. I had no intention to offend anyone. In my defense, in less than 3 days I got 5 different guys sending me almost letter for letter the same email/message about the same vulnerability when I normally get *maybe* one vulerability report in a year.

That made me wonder if it wasn't spam. Apparently it wasn't, for which I apologize.

**Comment by mvdkleijn** [comment]
to clarify: I didn't mean similar report structure, which is fine.

I meant an almost letter for letter copy of the earlier messages I got.

**Comment by mvdkleijn** [comment]
Fixed in 0.8.3

**Comment by oerdnj** [comment]
The issue is definitely not fixed. The fix that made it to 0.8.3 is unsufficient as there might be other extensions that could get executed either via PHP (f.e. default Debian config executes .phtml and .php5) or via some other interpreter.

You have to separate a code directories with user content directories and make sure that no user uploaded file can ever be executed.

```
+        $ext = strtolower(pathinfo($data['new_name'], PATHINFO_EXTENSION));
+        if (in_array($ext, ['php', 'php3', 'php4', 'inc')) {
+            Flash::set('error', __('Not allowed to upload files with extension :ext', $ext));
+            redirect(get_url('plugin/file_manager/browse/'));
+        }
```

This code is simply wrong as it protect against very specific class of attacks (and extensions) and still leaves the doors to attackers open via other means.

You should also make sure that you won't ever include a user uploaded page, f.e. via WOLFPAGE=<user_upload_file> (I am sorry, but I don't have a time for a complete audit and I don't care enough).
